### PR TITLE
Compatibility for gnome.extensions.org

### PR DESCRIPTION
**Pull request by schnz**
The latest version of gTile was rejected from being published on gnome.extensions.org.
See https://extensions.gnome.org/review/48580

This PR addresses the points raised by @jrahmatzadeh. I would appreciate if you could directly review to this PR, so we can skip unnecessary review-loops on EGO.

> 1. Don't use version when importing.

I am [still not convinced](https://github.com/oae/gnome-shell-pano/pull/222#issuecomment-1812880830) how this helps with type-safe development. But I gather from your [comment](https://github.com/oae/gnome-shell-pano/pull/222#issuecomment-1814075558) that you specifically refer to versioned imports that are also a direct dependency to `gnome-shell`. This only concerns the `Gio` and `Gtk` imports in this project. I've removed the version constrain from all of those imports in this PR.

> 2. Also null out on destroy (line 26 `prefs.js`):
>  ```js
>    this.#gc = null;
>    this.#settings = null;
>    this.#window = null;
>    ```

When I did that, I experienced a few errors while monitoring the logs from `/usr/bin/gjs` and `/usr/bin/gnome-shell`. I even left a comment about that in the [code base](https://github.com/gTile/gTile/blob/1ba3ac3951a2aa15867ae7f182e54abf9ed4585f/src/prefs.ts#L48-L53). Please have a look at it.

You can reproduce the errors by adding the lines you suggested. Then:
1. Make sure that `prefs.js` wasn't yet loaded by gnome-shell (so that the changes made take effect)
2. Open the preference dialog of the extension.
3. Set a shortcut for an arbitrary action (I believe this wasn't even required to trigger the errors)
4. Close the preference dialog
5. Now logout from the gnome session.
    - Note: I observed the logs remotely via SSH
6. You should see the errors that I referenced in the comment above.

> 3. Where do you remove the keybindings on disable in `core/HotkeyManager.js`?

In [`HotkeyManager::release()`](https://github.com/gTile/gTile/blob/1ba3ac3951a2aa15867ae7f182e54abf9ed4585f/src/core/HotkeyManager.ts#L97-L106).  Note the `this.#keyBindingGroupMask = 0` assignment. That causes all keybindings to be unregistered by the `register*Hotkeys()` calls that follow.

> 4. Timeouts should be removed on destroy:
>
>  - line 163 `ui/Overlay.js`
>  - line 14 `util/volatile.js`
>  - line 206 `core/OverlayManager.js`
>
>  [EGO Review Guidelines: Timeout](https://gjs.guide/extensions/review-guidelines/review-guidelines.html#remove-main-loop-sources)

The guideline you are referencing is using `GLib.timeout_add_seconds()` and `GLib.Source.remove()` in the example code. I was under the impression that a call to `setTimeout()` would merely require a call to `clearTimeout()` when destroying. Is that assumption correct? Or do I have to explicitly call `GLib.Source.remove`?

### CODE REVIEW

**Comment by jrahmatzadeh** [comment]
1. Still should remove the versions from all of those imports since they are getting specified by the parent module as mentioned in the other issue.

2. Tried on my main setup. No errors doing this in `prefs.js`:

    ```js
    #release() {
        this.#gc.release();
        this.#gc = null;
        this.#settings = null;
        this.#window = null;
    }
    ```

3. Removing keybindings by assigning `0` to `keyBindingGroupMask` is hard to follow but that's ok to pass the review since it works.

4. When you use `setTimeout()`, you can use `clearTimeout()` to clean up.


Anyway, since you mentioned about setting an action in prefs, setting any shortcut in actions giving an error like this (tested on v54):

``` 
JS ERROR: Error: 0x1c is not a valid value for flags argument modifiers
#onRebindKey/<@file:///home/ice/.local/share/gnome-shell/extensions/gTile@vibou/prefs.js:439:22
_init/GLib.MainLoop.prototype.runAsync/</<@resource:///org/gnome/gjs/modules/core/overrides/GLib.js:266:3
```

**Comment by scherepanov** [comment]
Let me know when it will be ready for merge.

**Comment by schnz** [comment]
> Still should remove the versions from all of those imports since they are getting specified by the parent module as mentioned in the other issue.

I am sorry. I still don't get it. What is the "parent module" can you point me to a reference where I can find what imports are referenced by it?

> Tried on my main setup. No errors doing this in prefs.js:

I tried to make a video of this but for some reason RDP doesn't work on my remote system.  However, I can very reliably produce these errors on _my_system. 

Now, since you also referenced another error that doesn't show up on my system: Mind to share your system details (the output of `hostnamectl` would be sufficient). Thanks.

**Comment by jrahmatzadeh** [comment]
> What is the "parent module"? Can you point me to a reference where I can find what imports are referenced by it?

- [meson](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/8d7dc098b158be0a9f980fed97cad9031287e642/meson.build)
- [dependencies](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/8d7dc098b158be0a9f980fed97cad9031287e642/js/misc/dependencies.js)
- [prefs](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/c4fcc82137b54a7cdcdaaa3b18d7582388365651/js/dbusServices/extensions/extensionPrefsDialog.js#L3-8)

> Mind to share your system details

Fedora 39, GNOME Shell 45.1 (X11).

**Comment by schnz** [comment]
> Fedora 39, GNOME Shell 45.1 (X11).

Also reproducible there. Are you sure that the `prefs.js` file wasn't cached/loaded by gnome-shell already before you applied the changes?

[Screencast from 2023-12-01 15-16-40.webm](https://github.com/gTile/gTile/assets/3457747/c1e4ed6d-9363-4359-a851-65b662cb1c67)

> JS ERROR: Error: 0x1c is not a valid value for flags argument modifiers

I'd be really interested in fixing this. But I merely forward pass-through the [state modifier](https://gjs-docs.gnome.org/gtk40~4.0/gtk.eventcontrollerkey#signal-key-pressed) from the `key-pressed` event to [`Gtk.accelerator_valid()`](https://gjs-docs.gnome.org/gtk40~4.0/gtk.accelerator_valid). Also the error doesn't show up on my system (nor on the Fedore live box). Can you tell me what modifier keys you pressed and which keyboard layout was in use?

**Comment by jrahmatzadeh** [comment]
https://github.com/gTile/gTile/assets/66198705/65c87d75-ddd3-4c23-9b68-55d80e27f4db

**Comment by jrahmatzadeh** [comment]
https://github.com/gTile/gTile/assets/66198705/f8978c1a-d78a-4a71-87cc-a88473522b1b

**Comment by scherepanov** [comment]
Is this PR ready for merging and publishing to EGO?

**Comment by jrahmatzadeh** [comment]
> Is this PR ready for merging and publishing to EGO?

No (as mentioned in my first reply).

**Comment by schnz** [comment]
> Tried on my main setup. No errors doing this in prefs.js:

How should we move forward with this? I really wouldn't want to introduce something that triggers errors on purpose and (seems) to be okay as is.

**Comment by jrahmatzadeh** [comment]
> How should we move forward with this? I really wouldn't want to introduce something that triggers errors on purpose and (seems) to be okay as is.

Remove line [420](https://github.com/gTile/gTile/blob/c37ecc601126b886b70ca637d7007983672824e4/src/prefs.ts#L420) and [482](https://github.com/gTile/gTile/blob/c37ecc601126b886b70ca637d7007983672824e4/src/prefs.ts#L482) (`prefs.js`) and try again.

**Comment by schnz** [comment]
> Remove line [420](https://github.com/gTile/gTile/blob/c37ecc601126b886b70ca637d7007983672824e4/src/prefs.ts#L420) and [482](https://github.com/gTile/gTile/blob/c37ecc601126b886b70ca637d7007983672824e4/src/prefs.ts#L482) (prefs.js) and try again.

Unfortunately, the errors remain. 

```diff
diff --git a/src/prefs.ts b/src/prefs.ts
index c3859b1..6ff7c07 100644
--- a/src/prefs.ts
+++ b/src/prefs.ts
@@ -45,6 +45,9 @@ export default class extends ExtensionPreferences {

   #release() {
     this.#gc.release();
+    this.#gc = undefined!;
+    this.#settings = undefined!;
+    this.#window = undefined!;
     // do NOT set #settings or other instance variables to undefined! This would
     // cause the GC to dereference and cleanup resources before GJS expects them
     // to be gone (even long after the preference window was closed). Would
@@ -417,7 +420,7 @@ export default class extends ExtensionPreferences {
         }
       }
     });
-    this.#gc.defer(() => this.#settings.disconnect(chid));
+    // this.#gc.defer(() => this.#settings.disconnect(chid));

     return row;
   }
@@ -479,7 +482,7 @@ const ShortcutRow = GObject.registerClass({
       label.label = this.#label();
       resetBtn.visible = this.#isCustomized();
     });
-    this.#gc.defer(() => this.#settings.disconnect(chid));
+    // this.#gc.defer(() => this.#settings.disconnect(chid));
   }

   release(): void {
```

**Comment by jrahmatzadeh** [comment]
Did you try to destroy the dialog on window close?

If that doesn't solve the issue, you can remove `transient_for` for this release and pass the ego review (since users waiting for 45 port). You can find a solution for that in the next release.

**Comment by scherepanov** [comment]
Hi @jrahmatzadeh 
Please advice, what we should do to make gTile good enough to pass EGO review.

**Comment by jrahmatzadeh** [comment]
@schnz
Nulling out those properties needed to pass the EGO review.

Anyway, if the `transient_for` finalizing the dialog after the window, then unset the `transient_for` and destroy the dialog on `ShortcutRow.release()`:

```js
this.#dialog?.set_transient_for(null);
this.#dialog?.destroy();
```

as mentioned in [the docs](https://gjs-docs.gnome.org/gtk40~4.0/gtk.window#method-set_transient_for):

> Passing null for parent unsets the current transient window.

About the event controller, afaics, the way you are handling the event controller, only works on Wayland not X11. I wrote [an example](https://gjs.guide/extensions/upgrading/gnome-shell-40.html#eventcontrollerkey) for gtk event controller before. maybe try to use that instead.

Feel free to ping me anytime, here or on [GNOME Matrix Channel](https://matrix.to/#/#extensions:gnome.org).

**Comment by JMurph2015** [comment]
FWIW I can also reproduce what @schnz is talking about, didn't even need try setting a hotkey.

I also don't see how null'ing out our references is a requirement to for review.  I checked https://gjs.guide/extensions/review-guidelines/review-guidelines.html#general-guidelines and while it references it in a snippet, it does not say that's a rule.  In C, that pattern is used to guard against dangling references, but that's not needed if we have a functioning GC.

**Comment by jrahmatzadeh** [comment]
> I also don't see how null'ing out our references is a requirement to for review.

It's one of the EGO review guidelines rules ([destroy all objects](https://gjs.guide/extensions/review-guidelines/review-guidelines.html#destroy-all-objects)).

**Comment by JMurph2015** [comment]
Setting references to null doesn't destroy anything.  Those objects still exist, that just sets our variable names to something else.  "Destroy all objects" refers specifically to objects that _have a `.destroy()` method_ and need to do things like unbind handlers and such.  Setting the references to null (in C) is to guard against dangling references so that you do not try to access memory that has been freed.  In GJS, the practice only makes sense to guard against objects accessing the window API after they have been de-initialized (and doing weird things).

**Comment by jrahmatzadeh** [comment]
If you don't null `#window`, you still see those errors?

`#gc` and `#settings` must be null out though. since they have been created inside `ExtensionPreferences`.

**Comment by JMurph2015** [comment]
@jrahmatzadeh I don't think you're following here.  _**Setting class variables to null does not destroy them**_

What that review guideline is saying is "if you create any objects (particularly GObjects) that have a `.destroy()` or `.release()` or other finalizer method), you *must* call them" so that they will unregister their signal handlers and such.  That is the first line of that code snippet's method.  That is the only concept of "destroying" that makes sense in a garbage collected language.  The nulling is a style thing that replaces one category of bugs with another.  It doesn't do any destroying.

**Comment by schnz** [comment]
> then unset the `transient_for` [...] as mentioned in [the docs](https://gjs-docs.gnome.org/gtk40~4.0/gtk.window#method-set_transient_for):
> 
> > Passing null for parent unsets the current transient window.

That is indeed a workaround I can live with. Thanks for the tip. 

> Feel free to ping me anytime, here or on [GNOME Matrix Channel](https://matrix.to/#/#extensions:gnome.org).

Good to know! Thank you!

---

_Note to self: Next time, read the whole conversation before starting to implement something_

Following the conversation of you two, I see that #window does not need to be set to `null` after all. That's even better than the workaround with `set_transient_for(null)`. I wouldn't mind setting my own instance variables to null. That doesn't cause any issues after all.

> About the event controller, afaics, the way you are handling the event controller, only works on Wayland not X11. I wrote [an example](https://gjs.guide/extensions/upgrading/gnome-shell-40.html#eventcontrollerkey) for gtk event controller before. maybe try to use that instead.

That would explain why the dialog isn't working for you. I implemented it similar to the native Gnome shortcut settings dialog (see [here](https://gitlab.gnome.org/GNOME/gnome-control-center/-/blob/6228caff69b9d83e79d5d22d1015af568e7c3cd4/panels/keyboard/cc-keyboard-shortcut-editor.ui#L276-279) and [here](https://gitlab.gnome.org/GNOME/gnome-control-center/-/blob/6228caff69b9d83e79d5d22d1015af568e7c3cd4/panels/keyboard/keyboard-shortcuts.c#L280-287)) and hoped that it would work. I tested it with X11 right now but it continues to work as expected. The example of yours seems to be related to the legacy way of creating a preference window (prior to Gnome 44 and with Gtk 3). But as far as I can tell, the current implementation pretty much translates to the example of yours.

**Can anyone verify that the shortcut dialog isn't (or is) working on X11?**

**Comment by schnz** [comment]
Summary:

* The current PR addresses and resolves all issues raised in the EGO review
* In theory: It is ready to be merged (probably after a thumbs-up from @jrahmatzadeh)
* **But**: There seems to be a bug, as seen in [the video](https://github.com/gTile/gTile/pull/356#issuecomment-1836763561) above that the "set shortcut" dialog window is not working properly. The working theory is that it is related to X11, yet I cant reproduce it on my system(s).

I see two paths forward:
* Someone is able to reproduce the bug and provides the missing detail that causes it. Then I will also address it in this PR.
* Nobody is able/willing to reproduce it, say within the next 2 days. Then I propose to merge the PR. If the bug happens for other users systems we will likely see an issue being opened relatively fast.

FYI @scherepanov , @JMurph2015

**Comment by scherepanov** [comment]
Great job, @schnz !
I cannot test this pr - I am away from home on vacations.
If it would be acceptable, we can merge this PR as-is - it does contain set of important fixes.
We can open new PR specifically for @jrahmatzadeh review.
Let me know if this is acceptable.

**Comment by jrahmatzadeh** [comment]
LGTM. Please post to EGO for the review.

**Comment by schnz** [comment]
Thanks for your patience, jrahmatzadeh. 

Just merged & released the new version & double-checked that no errors are occurring. Ready to be submitted by you, @scherepanov.

**Comment by scherepanov** [comment]
New version posted to EGO.
Thanks @jrahmatzadeh for your help with code!
